stdlib/public/core/Diffing.swift

```diff
@@ -0,0 +1,782 @@
+  /// Returns the difference needed to produce the receiver's state from the
+  /// parameter's state, using equality to establish equivalence between
+  /// elements.
+  ///
+  /// This function does not infer element moves, but they can be computed
+  /// using `CollectionDifference.inferringMoves()` if desired.
+  ///
+  /// - Parameters:
+  ///   - other: The base state.
+  ///
+  /// - Returns: The difference needed to produce the reciever's state from
+  ///   the parameter's state.
+  ///
+  /// - Complexity: For pathological inputs, worst case performance is
+  ///   O(`self.count` * `other.count`). Faster execution can be expected
+  ///   when the collections share many common elements, or if `Element`
+  ///   also conforms to `Hashable`.
+  public func difference<C>(from other: C) -> CollectionDifference<Element>
+    where C: BidirectionalCollection, C.Element == Self.Element
+  {
+    return difference(from: other, by: ==)
+  }
+}
+
+extension BidirectionalCollection {
+  /// Returns a pair of subsequences containing the initial elements that
+  /// `self` and `other` have in common.
+  fileprivate func commonPrefix<Other : BidirectionalCollection>(
+    with other: Other, by areEquivalent: (Element, Other.Element) -> Bool
+  ) -> (SubSequence, Other.SubSequence) where Element == Other.Element {
+    let (s1, s2) = (startIndex, other.startIndex)
+    let (e1, e2) = (endIndex, other.endIndex)
+    var (i1, i2) = (s1, s2)
+    while i1 != e1 && i2 != e2 {
+      if !areEquivalent(self[i1], other[i2]) { break }
+      formIndex(after: &i1)
+      other.formIndex(after: &i2)
+    }
+    return (self[s1..<i1], other[s2..<i2])
+  }
+}
+
+// MARK: Diff production from OrderedCollection (soon to be BidirectionalCollection)
+
+/// A collection of changes between a source and target collection.
+///
+/// It can be used to traverse the [longest common subsequence][lcs] of
+/// source and target:
+///
+///   let changes = CollectionChanges(from: source, to target)
+///   for case let .match(s, t) in changes {
+///     // use `s`, `t`
+///   }
+///
+/// It can also be used to traverse the [shortest edit script][ses] of
+/// remove and insert operations:
+///
+///   let changes = CollectionChanges(from: source, to target)
+///   for c in changes {
+///     switch c {
+///     case let .removed(s):
+///       // use `s`
+///     case let .inserted(t):
+///       // use `t`
+///     case .matched: continue
+///     }
+///   }
+///
+/// [lcs]: http://en.wikipedia.org/wiki/Longest_common_subsequence_problem
+/// [ses]: http://en.wikipedia.org/wiki/Edit_distance
+///
+/// - Note: `CollectionChanges` holds a reference to state used to run the
+///     difference algorithm, which can be exponentially larger than the
+///     changes themselves.
+fileprivate struct CollectionChanges<
+  SourceIndex : Comparable, TargetIndex : Comparable
+> {
+  typealias Endpoint = (x: SourceIndex, y: TargetIndex)
```

Comment:
We'll need an explicit access level for the typealias, too:
```suggestion
  private typealias Endpoint = (x: SourceIndex, y: TargetIndex)
```